File: python/railjson_generator/railjson_generator/schema/infra/neutral_section.py
from dataclasses import dataclass, field
from typing import List

from osrd_schemas import infra

from railjson_generator.schema.infra.direction import Direction
from railjson_generator.schema.infra.range_elements import DirectionalTrackRange
from railjson_generator.schema.infra.track_section import TrackSection


def _neutral_section_id():
    res = f"neutral_section.{NeutralSection._index}"
    NeutralSection._index += 1
    return res


@dataclass
class NeutralSection:
    """
    Neutral sections are portions of track where trains aren't allowed to pull power from electrifications.
    They have to rely on inertia to cross such sections.

    In practice, neutral sections are delimited by signs. In OSRD, neutral sections are directional
    to allow accounting for different sign placement depending on the direction.

    For more details see [the documentation](https://osrd.fr/en/docs/explanation/neutral_sections/).
    """

    announcement_track_ranges: List[DirectionalTrackRange] = field(default_factory=list)
    track_ranges: List[DirectionalTrackRange] = field(default_factory=list)
    lower_pantograph: bool = field(default=False)
    label: str = field(default_factory=_neutral_section_id)

    _index = 0

    @staticmethod
    def reset_index():
        NeutralSection._index = 0

    def add_track_range(
        self, track: TrackSection, begin: float, end: float, direction: Direction
    ):
        self.track_ranges.append(
            DirectionalTrackRange(
                begin=begin,
                end=end,
                track=track,
                direction=direction,
            )
        )

    def add_announcement_track_range(
        self, track: TrackSection, begin: float, end: float, direction: Direction
    ):
        self.announcement_track_ranges.append(
            DirectionalTrackRange(
                begin=begin,
                end=end,
                track=track,
                direction=direction,
            )
        )

    def to_rjs(self):
        return infra.NeutralSection(
            id=self.label,
            announcement_track_ranges=[
                track.to_rjs() for track in self.announcement_track_ranges
            ],
            track_ranges=[track.to_rjs() for track in self.track_ranges],
            lower_pantograph=self.lower_pantograph,
        )

File: editoast/schemas/src/infra/neutral_section.rs
use serde::Deserialize;
use serde::Serialize;
use utoipa::ToSchema;

use super::DirectionalTrackRange;
use super::Sign;
use crate::primitives::Identifier;
use crate::primitives::OSRDIdentified;
use crate::primitives::OSRDTyped;
use crate::primitives::ObjectType;

/// Neutral sections are portions of track where trains aren't allowed to pull power from electrifications. They have to rely on inertia to cross such sections.
///
/// In practice, neutral sections are delimited by signs. In OSRD, neutral sections are directional to allow accounting for different sign placement depending on the direction.
///
/// For more details see [the documentation](https://osrd.fr/en/docs/explanation/neutral_sections/).
#[derive(Debug, Default, Clone, Deserialize, Serialize, PartialEq, ToSchema)]
#[serde(deny_unknown_fields)]
pub struct NeutralSection {
    #[schema(inline)]
    pub id: Identifier,
    pub announcement_track_ranges: Vec<DirectionalTrackRange>,
    pub track_ranges: Vec<DirectionalTrackRange>,
    pub lower_pantograph: bool, // Whether the trains need to lower their pantograph to cross this section
    #[serde(default)]
    #[schema(inline)]
    pub extensions: NeutralSectionExtensions,
}

#[derive(Debug, Default, Clone, Deserialize, Serialize, PartialEq, ToSchema)]
#[serde(deny_unknown_fields)]
pub struct NeutralSectionExtensions {
    #[schema(inline)]
    pub neutral_sncf: Option<NeutralSectionNeutralSncfExtension>,
}

#[derive(Debug, Default, Clone, Deserialize, Serialize, PartialEq, ToSchema)]
#[serde(deny_unknown_fields)]
pub struct NeutralSectionNeutralSncfExtension {
    pub announcement: Vec<Sign>,
    pub exe: Sign,
    pub end: Vec<Sign>,
    pub rev: Vec<Sign>,
}

impl OSRDTyped for NeutralSection {
    fn get_type() -> ObjectType {
        ObjectType::NeutralSection
    }
}

impl OSRDIdentified for NeutralSection {
    fn get_id(&self) -> &String {
        &self.id
    }
}
